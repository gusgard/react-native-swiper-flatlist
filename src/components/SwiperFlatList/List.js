import React from 'react';

import SwiperFlatList from './SwiperFlatList';

// TODO rename to index.
export default class List extends React.PureComponent {
  constructor(props) {
    super(props);

    this.viewabilityConfig = {
      // waitForInteraction: true,
      // viewAreaCoveragePercentThreshold: 95,
      minimumViewTime: 1000,
      itemVisiblePercentThreshold: 100,
    };
  }

  render() {
    // return <SwiperFlatList {...this.props} test={this.viewabilityConfig}></SwiperFlatList>;
    // console.log(this.flatListRef);

    return (
      <SwiperFlatList
        {...this.props}
        // ref={component => {
        //   this.flatListRef = component;
        // }}
        viewabilityConfig={this.viewabilityConfig}
        onViewableItemsChanged={({ viewableItems, changed }) => {
          console.log('viewableItems', viewableItems?.[0]?.index);
          console.log('changed', changed?.[0]?.index);
        }}
      />
    );
  }
}
