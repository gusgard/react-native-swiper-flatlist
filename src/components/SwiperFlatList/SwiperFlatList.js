import React from 'react';
import PropTypes from 'prop-types';
import { FlatList } from 'react-native';

import Pagination from '../Pagination';

const MILLISECONDS = 1000;
const FIRST_INDEX = 0;
const ITEM_VISIBLE_PERCENT_THRESHOLD = 60;

const SwiperFlatList = React.forwardRef(
  (
    {
      vertical,
      children,
      data,
      renderItem,
      renderAll,
      index,
      // Pagination
      showPagination,
      PaginationComponent,
      paginationActiveColor,
      paginationDefaultColor,
      paginationStyle,
      paginationStyleItem,
      paginationStyleItemActive,
      paginationStyleItemInactive,
      onPaginationSelectedIndex,
      // Autoplay
      autoplayDelay,
      autoplay,
      autoplayLoop,
      autoplayLoopKeepAnimation,
      autoplayInvertDirection,
      // Functions
      onChangeIndex,
      onMomentumScrollEnd,
      onViewableItemsChanged,
      viewabilityConfig,
      disableGesture,
      e2eId,
      ...props
    },
    ref,
  ) => {
    let _data;
    let _renderItem;

    if (children) {
      // github.com/gusgard/react-native-swiper-flatlist/issues/40
      _data = Array.isArray(children) ? children : [children];
      _renderItem = ({ item }) => item;
    } else if (data) {
      _data = data;
      _renderItem = renderItem;
    }
    const size = _data.length;
    // Items to render in the initial batch.
    const _initialNumToRender = renderAll ? size : 1;
    const [paginationIndex, setPaginationIndex] = React.useState(index);
    const [prevIndex, setPrevIndex] = React.useState(index);
    const [paginationIndexes, setPaginationIndexes] = React.useState({ index, prevIndex: index });
    const [ignoreOnMomentumScrollEnd, setIgnoreOnMomentumScrollEnd] = React.useState(false);
    const flatListElement = React.useRef(null);
    const [scrollEnabled, setScrollEnabled] = React.useState(!disableGesture);

    const _onChangeIndex = React.useCallback(
      ({ index: _index, prevIndex: _prevIndex }) => {
        onChangeIndex?.({ index: _index, prevIndex: _prevIndex });
      },
      [onChangeIndex],
    );

    const _scrollToIndex = params => {
      if (typeof params !== 'object') {
        console.error(
          'Expected an object for "scrollToIndex", for example: scrollToIndex({ index: 1, animated: true })',
        );
        // NOTE: remove in future versions.
        return;
      }

      const { index: indexToScroll, animated = true } = params;
      const newParams = { animated, index: indexToScroll };

      setPaginationIndexes(prevState => {
        setIgnoreOnMomentumScrollEnd(true);
        return { index: indexToScroll, prevIndex: prevState.index };
      });
      // When execute "scrollToIndex", we ignore the method "onMomentumScrollEnd"
      // because it not working on Android
      // https://github.com/facebook/react-native/issues/21718
      flatListElement?.current?.scrollToIndex(newParams);
    };

    React.useEffect(() => {
      const next = {
        index: paginationIndexes.index,
        prevIndex: paginationIndexes.prevIndex,
      };
      if (paginationIndex !== next.index) {
        setPaginationIndex(next.index);
      }
      if (prevIndex !== next.prevIndex) {
        setPrevIndex(next.prevIndex);
      }
      _onChangeIndex({ index: next.index, prevIndex: next.prevIndex });
      // only consider "paginationIndexes"
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [paginationIndexes]);

    React.useEffect(() => {
      _onChangeIndex({ index: paginationIndex, prevIndex: prevIndex });
    }, [paginationIndex, prevIndex]);

    React.useImperativeHandle(ref, () => ({
      scrollToIndex: (...args) => {
        setScrollEnabled(true);
        _scrollToIndex(...args);
        setScrollEnabled(!disableGesture);
      },
      getCurrentIndex: () => paginationIndex,
      getPrevIndex: () => prevIndex,
      goToLastIndex: () => {
        setScrollEnabled(true);
        _scrollToIndex({ index: size - 1 });
        setScrollEnabled(!disableGesture);
      },
      goToFirstIndex: () => {
        setScrollEnabled(true);
        _scrollToIndex({ index: FIRST_INDEX });
        setScrollEnabled(!disableGesture);
      },
    }));

    React.useEffect(() => {
      const isLastIndexEnd = autoplayInvertDirection
        ? paginationIndex === FIRST_INDEX
        : paginationIndex === _data.length - 1;
      const shouldContinuoWithAutoplay = autoplay && !isLastIndexEnd;
      let autoplayTimer;
      if (shouldContinuoWithAutoplay || autoplayLoop) {
        autoplayTimer = setTimeout(() => {
          if (_data.length < 1) {
            // avoid nextIndex being set to NaN
            return;
          }

          const nextIncrement = autoplayInvertDirection ? -1 : +1;

          let nextIndex = (paginationIndex + nextIncrement) % _data.length;
          if (autoplayInvertDirection && nextIndex < FIRST_INDEX) {
            nextIndex = _data.length - 1;
          }

          // Disable end loop animation unless `autoplayLoopKeepAnimation` prop configured
          const animate = !isLastIndexEnd || autoplayLoopKeepAnimation;

          _scrollToIndex({ index: nextIndex, animated: animate });
        }, autoplayDelay * MILLISECONDS);
      }
      // https://upmostly.com/tutorials/settimeout-in-react-components-using-hooks
      return () => clearTimeout(autoplayTimer);
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [paginationIndex]);
    const _onMomentumScrollEnd = e => {
      // NOTE: Method not executed when call "flatListElement?.current?.scrollToIndex"
      if (ignoreOnMomentumScrollEnd) {
        setIgnoreOnMomentumScrollEnd(false);
        return;
      }

      onMomentumScrollEnd?.({ index: paginationIndex }, e);

      //_onChangeIndex({ index: paginationIndex, prevIndex });
    };

    const _onViewableItemsChanged = React.useMemo(
      () => params => {
        const { changed } = params;
        const newItem = changed?.[FIRST_INDEX];
        if (newItem !== undefined) {
          const nextIndex = newItem.index;
          if (newItem.isViewable) {
            setPaginationIndex(nextIndex);
          } else {
            setPrevIndex(nextIndex);
          }
        }
        onViewableItemsChanged?.(params);
      },
      [],
    );

    const flatListProps = {
      scrollEnabled,
      ref: flatListElement,
      keyExtractor: (_item, _index) => _index.toString(),
      horizontal: !vertical,
      showsHorizontalScrollIndicator: false,
      showsVerticalScrollIndicator: false,
      pagingEnabled: true,
      ...props,
      onMomentumScrollEnd: _onMomentumScrollEnd,
      onScrollToIndexFailed: info =>
        setTimeout(() => _scrollToIndex({ index: info.index, animated: false })),
      data: _data,
      renderItem: _renderItem,
      initialNumToRender: _initialNumToRender,
      initialScrollIndex: index, // used with onScrollToIndexFailed
      viewabilityConfig: {
        // https://facebook.github.io/react-native/docs/flatlist#minimumviewtime
        minimumViewTime: 200,
        itemVisiblePercentThreshold: ITEM_VISIBLE_PERCENT_THRESHOLD,
        ...viewabilityConfig,
      },
      onViewableItemsChanged: _onViewableItemsChanged,
      // debug: true, // for debug
      testID: e2eId,
    };

    const paginationProps = {
      size,
      paginationIndex: paginationIndex,
      scrollToIndex: _scrollToIndex,
      paginationActiveColor,
      paginationDefaultColor,
      paginationStyle,
      paginationStyleItem,
      paginationStyleItemActive,
      paginationStyleItemInactive,
      onPaginationSelectedIndex,
      e2eId,
    };

    return (
      <React.Fragment>
        <FlatList {...flatListProps} />
        {showPagination && <PaginationComponent {...paginationProps} />}
      </React.Fragment>
    );
  },
);

SwiperFlatList.propTypes = {
  data: PropTypes.array,
  vertical: PropTypes.bool,
  index: PropTypes.number,
  renderAll: PropTypes.bool,
  renderItem: PropTypes.func,
  // Only is allowed children or data not both
  children(props, propName) {
    const { data } = props;
    if (!props[propName] && !data) {
      return new Error('Invalid props, `data` or `children` is required');
    }
    if (data && data.length !== 0 && !props.renderItem) {
      return new Error('Invalid props, `renderItem` is required');
    }
    return undefined;
  },
  onChangeIndex: PropTypes.func,

  // Pagination
  showPagination: PropTypes.bool,
  PaginationComponent: PropTypes.func,
  paginationActiveColor: Pagination.propTypes.paginationActiveColor,
  paginationDefaultColor: Pagination.propTypes.paginationDefaultColor,
  paginationStyle: Pagination.propTypes.paginationStyle,
  paginationStyleItem: Pagination.propTypes.paginationStyleItem,
  paginationStyleItemActive: Pagination.propTypes.paginationStyleItemActive,
  paginationStyleItemInactive: Pagination.propTypes.paginationStyleItemInactive,
  onPaginationSelectedIndex: Pagination.propTypes.onPaginationSelectedIndex,

  // Autoplay
  autoplayDelay: PropTypes.number,
  autoplay: PropTypes.bool,
  autoplayInvertDirection: PropTypes.bool,
  autoplayLoop: PropTypes.bool,
  autoplayLoopKeepAnimation: PropTypes.bool,

  // Optionals
  onMomentumScrollEnd: PropTypes.func,
  onViewableItemsChanged: PropTypes.func,
  viewabilityConfig: PropTypes.object,
  disableGesture: PropTypes.bool,
  e2eId: PropTypes.string,
};

SwiperFlatList.defaultProps = {
  index: FIRST_INDEX,
  data: [],
  autoplayDelay: 3,
  autoplayInvertDirection: false,
  autoplayLoop: false,
  autoplayLoopKeepAnimation: false,
  autoplay: false,
  showPagination: false,
  vertical: false,
  renderAll: false,
  PaginationComponent: Pagination,
  onChangeIndex: undefined,
  // Optionals
  onMomentumScrollEnd: undefined,
  onViewableItemsChanged: undefined,
  viewabilityConfig: {},
  disableGesture: false,
  e2eId: undefined,
};

export default SwiperFlatList;
