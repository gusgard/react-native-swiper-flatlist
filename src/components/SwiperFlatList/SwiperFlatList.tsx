import React from 'react';
import { FlatList, FlatListProps, Platform, Dimensions } from 'react-native';

import { Pagination } from '../Pagination/Pagination';
import { SwiperFlatListProps, SwiperFlatListRefProps } from './SwiperFlatListProps';

const MILLISECONDS = 1000;
const FIRST_INDEX = 0;
const ITEM_VISIBLE_PERCENT_THRESHOLD = 60;

// TODO: figure out how to use forwardRef with generics
type T1 = any;
type ScrollToIndex = { index: number; animated?: boolean };
type ScrollToIndexInternal = { useOnChangeIndex: boolean };

const { width: WIDTH, height: HEIGHT } = Dimensions.get('window');

// const SwiperFlatList = React.forwardRef<RefProps, SwiperFlatListProps<SwiperType>>(
export const SwiperFlatList = React.forwardRef(
  // <T1 extends any>(
  (
    {
      vertical = false,
      children,
      data = [],
      renderItem,
      renderAll = false,
      index = FIRST_INDEX,
      // Pagination
      showPagination = false,
      PaginationComponent = Pagination,
      paginationActiveColor,
      paginationDefaultColor,
      paginationStyle,
      paginationStyleItem,
      paginationStyleItemActive,
      paginationStyleItemInactive,
      onPaginationSelectedIndex,
      // Autoplay
      autoplayDelay = 3,
      autoplay = false,
      autoplayLoop = false,
      autoplayLoopKeepAnimation = false,
      autoplayInvertDirection = false,
      // Functions
      onChangeIndex,
      onMomentumScrollEnd,
      onViewableItemsChanged,
      viewabilityConfig = {},
      disableGesture = false,
      e2eID,
      ...props
    }: SwiperFlatListProps<T1>,
    ref: React.Ref<SwiperFlatListRefProps>,
  ) => {
    let _data: unknown[] = [];
    let _renderItem: FlatListProps<any>['renderItem'];

    if (children) {
      // github.com/gusgard/react-native-swiper-flatlist/issues/40
      _data = Array.isArray(children) ? children : [children];
      _renderItem = ({ item }) => item;
    } else if (data) {
      _data = data;
      _renderItem = renderItem;
    } else {
      console.error('Invalid props, `data` or `children` is required');
    }
    const size = _data.length;
    // Items to render in the initial batch.
    const _initialNumToRender = renderAll ? size : 1;
    const [currentIndexes, setCurrentIndexes] = React.useState({ index, prevIndex: index });
    const [ignoreOnMomentumScrollEnd, setIgnoreOnMomentumScrollEnd] = React.useState(false);
    const flatListElement = React.useRef<FlatList<unknown>>(null);
    const [scrollEnabled, setScrollEnabled] = React.useState(!disableGesture);

    const _onChangeIndex = React.useCallback(
      ({ index: _index, prevIndex: _prevIndex }) => {
        if (_index !== _prevIndex) {
          onChangeIndex?.({ index: _index, prevIndex: _prevIndex });
        }
      },
      [onChangeIndex],
    );

    const _scrollToIndex = (params: ScrollToIndex, extra: ScrollToIndexInternal) => {
      const { index: indexToScroll, animated = true } = params;
      const newParams = { animated, index: indexToScroll };

      setIgnoreOnMomentumScrollEnd(true);

      const next = {
        index: indexToScroll,
        prevIndex: currentIndexes.index,
      };
      if (currentIndexes.index !== next.index && currentIndexes.prevIndex !== next.prevIndex) {
        setCurrentIndexes({ index: next.index, prevIndex: next.prevIndex });
      } else if (currentIndexes.index !== next.index) {
        setCurrentIndexes((prevState) => ({ ...prevState, index: next.index }));
      } else if (currentIndexes.prevIndex !== next.prevIndex) {
        setCurrentIndexes((prevState) => ({ ...prevState, prevIndex: next.prevIndex }));
      }

      if (extra.useOnChangeIndex) {
        _onChangeIndex({ index: next.index, prevIndex: next.prevIndex });
      }

      // When execute "scrollToIndex", we ignore the method "onMomentumScrollEnd"
      // because it not working on Android
      // https://github.com/facebook/react-native/issues/21718
      flatListElement?.current?.scrollToIndex(newParams);
    };

    // change the index when the user swipe the items
    React.useEffect(() => {
      _onChangeIndex({ index: currentIndexes.index, prevIndex: currentIndexes.prevIndex });
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [currentIndexes.index]);

    React.useImperativeHandle(ref, () => ({
      scrollToIndex: (item: ScrollToIndex) => {
        setScrollEnabled(true);
        _scrollToIndex(item, { useOnChangeIndex: true });
        setScrollEnabled(!disableGesture);
      },
      getCurrentIndex: () => currentIndexes.index,
      getPrevIndex: () => currentIndexes.prevIndex,
      goToLastIndex: () => {
        setScrollEnabled(true);
        _scrollToIndex({ index: size - 1 }, { useOnChangeIndex: false });
        setScrollEnabled(!disableGesture);
      },
      goToFirstIndex: () => {
        setScrollEnabled(true);
        _scrollToIndex({ index: FIRST_INDEX }, { useOnChangeIndex: false });
        setScrollEnabled(!disableGesture);
      },
    }));

    React.useEffect(() => {
      const isLastIndexEnd = autoplayInvertDirection
        ? currentIndexes.index === FIRST_INDEX
        : currentIndexes.index === _data.length - 1;
      const shouldContinuoWithAutoplay = autoplay && !isLastIndexEnd;
      let autoplayTimer: ReturnType<typeof setTimeout>;
      if (shouldContinuoWithAutoplay || autoplayLoop) {
        autoplayTimer = setTimeout(() => {
          if (_data.length < 1) {
            // avoid nextIndex being set to NaN
            return;
          }

          const nextIncrement = autoplayInvertDirection ? -1 : +1;

          let nextIndex = (currentIndexes.index + nextIncrement) % _data.length;
          if (autoplayInvertDirection && nextIndex < FIRST_INDEX) {
            nextIndex = _data.length - 1;
          }

          // Disable end loop animation unless `autoplayLoopKeepAnimation` prop configured
          const animate = !isLastIndexEnd || autoplayLoopKeepAnimation;

          _scrollToIndex({ index: nextIndex, animated: animate }, { useOnChangeIndex: true });
        }, autoplayDelay * MILLISECONDS);
      }
      // https://upmostly.com/tutorials/settimeout-in-react-components-using-hooks
      return () => clearTimeout(autoplayTimer);
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [currentIndexes.index, _data.length]);

    const _onMomentumScrollEnd: FlatListProps<unknown>['onMomentumScrollEnd'] = (event) => {
      // NOTE: Method not executed when call "flatListElement?.current?.scrollToIndex"
      if (ignoreOnMomentumScrollEnd) {
        setIgnoreOnMomentumScrollEnd(false);
        return;
      }

      onMomentumScrollEnd?.({ index: currentIndexes.index }, event);
    };

    const _onViewableItemsChanged = React.useMemo<FlatListProps<unknown>['onViewableItemsChanged']>(
      () => (params) => {
        const { changed } = params;
        const newItem = changed?.[FIRST_INDEX];
        if (newItem !== undefined) {
          const nextIndex = newItem.index as number;
          if (newItem.isViewable) {
            setCurrentIndexes((prevState) => ({ ...prevState, index: nextIndex }));
          } else {
            setCurrentIndexes((prevState) => ({ ...prevState, prevIndex: nextIndex }));
          }
        }
        onViewableItemsChanged?.(params);
      },
      // eslint-disable-next-line react-hooks/exhaustive-deps
      [],
    );

    const keyExtractor: FlatListProps<unknown>['keyExtractor'] = (_item, _index) =>
      _index.toString();
    const onScrollToIndexFailed: FlatListProps<unknown>['onScrollToIndexFailed'] = (info) =>
      setTimeout(() =>
        _scrollToIndex({ index: info.index, animated: false }, { useOnChangeIndex: true }),
      );

    const flatListProps = {
      scrollEnabled,
      ref: flatListElement,
      keyExtractor,
      horizontal: !vertical,
      showsHorizontalScrollIndicator: false,
      showsVerticalScrollIndicator: false,
      pagingEnabled: true,
      ...props,
      onMomentumScrollEnd: _onMomentumScrollEnd,
      onScrollToIndexFailed: onScrollToIndexFailed,
      data: _data,
      renderItem: _renderItem,
      initialNumToRender: _initialNumToRender,
      initialScrollIndex: index, // used with onScrollToIndexFailed
      viewabilityConfig: {
        // https://facebook.github.io/react-native/docs/flatlist#minimumviewtime
        minimumViewTime: 200,
        itemVisiblePercentThreshold: ITEM_VISIBLE_PERCENT_THRESHOLD,
        ...viewabilityConfig,
      },
      onViewableItemsChanged: _onViewableItemsChanged,
      // debug: true, // for debug
      testID: e2eID,
    };

    if (Platform.OS === 'web') {
      if (props.getItemLayout === undefined) {
        // NOTE: should we pass height/width for getItemLayout?
        const ITEM_DIMENSION = vertical ? HEIGHT : WIDTH;
        flatListProps.getItemLayout = (__data, ItemIndex: number) => {
          return {
            length: ITEM_DIMENSION,
            offset: ITEM_DIMENSION * ItemIndex,
            index: ItemIndex,
          };
        };
      }

      (flatListProps as any).dataSet = { 'paging-enabled-fix': true };
    }

    const scrollToIndexForPagination = (params: ScrollToIndex) => {
      _scrollToIndex(params, { useOnChangeIndex: false });
    };

    const paginationProps = {
      size,
      paginationIndex: currentIndexes.index,
      scrollToIndex: scrollToIndexForPagination,
      paginationActiveColor,
      paginationDefaultColor,
      paginationStyle,
      paginationStyleItem,
      paginationStyleItemActive,
      paginationStyleItemInactive,
      onPaginationSelectedIndex,
      e2eID,
    };

    return (
      <React.Fragment>
        <FlatList {...flatListProps} />
        {showPagination && <PaginationComponent {...paginationProps} />}
      </React.Fragment>
    );
  },
);

// https://gist.github.com/Venryx/7cff24b17867da305fff12c6f8ef6f96
type Handle<T> = T extends React.ForwardRefExoticComponent<React.RefAttributes<infer T2>>
  ? T2
  : never;
export type SwiperFlatList = Handle<typeof SwiperFlatList>;
