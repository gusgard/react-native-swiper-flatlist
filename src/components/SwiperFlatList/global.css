[data-paging-enabled-fix] > div > div > div {
    height: 100%;
}
