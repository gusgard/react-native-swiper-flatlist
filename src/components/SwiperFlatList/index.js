import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { View, FlatList } from 'react-native';

import Pagination from '../Pagination';

export default class SwiperFlatList extends PureComponent {
  static propTypes = {
    data: PropTypes.array.isRequired,
    onMomentumScrollEnd: PropTypes.func,
    vertical: PropTypes.bool.isRequired,
    index: PropTypes.number.isRequired,

    showPagination: PropTypes.bool.isRequired,
    paginationActiveColor: PropTypes.string,
    paginationDefaultColor: PropTypes.string,

    autoplayDelay: PropTypes.number.isRequired,
    autoplay: PropTypes.bool.isRequired,
    autoplayDirection: PropTypes.bool.isRequired,
    autoplayLoop: PropTypes.bool.isRequired,

    renderAll: PropTypes.bool,
    renderItem: PropTypes.func,

    PaginationComponent: PropTypes.func,

    // Only is allowed children or data not both
    children(props, propName) {
      const { data } = props;
      if (!props[propName] && !data) {
        return new Error('Invalid props, `data` or `children` is required');
      }
      if (data && data.length !== 0 && !props.renderItem) {
        return new Error('Invalid props, `renderItem` is required');
      }
      return undefined;
    }
  };

  static defaultProps = {
    index: 0,
    data: [],
    autoplayDelay: 3,
    autoplayDirection: true,
    autoplayLoop: false,
    autoplay: false,
    showPagination: false,
    vertical: false,
    renderAll: false,
    PaginationComponent: Pagination
  };

  componentWillMount() {
    this.setup(this.props);
    this.setState({ paginationIndex: this.props.index });
  }
  componentDidMount() {
    const { autoplay, index } = this.props;
    if (autoplay) {
      this._autoplay(index);
    }

    if (index !== 0) {
      this._scrollToIndex(index, false);
    }
  }

  componentWillUpdate(nextProps) {
    this.setup(nextProps);
  }

  componentWillUnmount() {
    if (this.autoplayTimer) {
      clearTimeout(this.autoplayTimer);
    }
  }

  setup = props => {
    const { children, data, renderItem, renderAll } = props;
    if (children) {
      this._data = children;
      this._renderItem = this.renderChildren;
    } else if (data) {
      this._data = data;
      this._renderItem = renderItem;
    }
    // Items to render in the initial batch.
    this._initialNumToRender = renderAll ? this._data.length : 1;
  };

  _autoplay = index => {
    const { autoplayDelay, autoplayLoop } = this.props;
    if (this.autoplayTimer) {
      clearTimeout(this.autoplayTimer);
    }
    const isEnd = index === this._data.length - 1;

    if (autoplayLoop || !isEnd) {
      this.autoplayTimer = setTimeout(() => {
        const nextIndex = (index + 1) % this._data.length;
        this._scrollToIndex(nextIndex, !isEnd);
      }, autoplayDelay * 1000);
    }
    if (isEnd) {
      // When scroll to the end and animated is false need to restart the autoplay
      setTimeout(() => {
        this.autoplayTimer = setTimeout(
          () => this._scrollToIndex(1, true),
          autoplayDelay * 1000
        );
      }, autoplayDelay * 1000);
    }
  };

  _scrollToIndex = (index, animated = true) => {
    const params = { animated, index };
    this.setState(() => {
      if (this.flatListRef) {
        this.flatListRef.scrollToIndex(params);
      }
      return { paginationIndex: index };
    });
  };

  _onMomentumScrollEnd = e => {
    const { autoplay, vertical, onMomentumScrollEnd } = this.props;
    const { contentOffset, layoutMeasurement } = e.nativeEvent;
    let index;
    if (vertical) {
      index = Math.floor(contentOffset.y / layoutMeasurement.height);
    } else {
      // Divide the horizontal offset by the width of the view to see which page is visible
      index = Math.floor(contentOffset.x / layoutMeasurement.width);
    }

    if (autoplay) {
      this._autoplay(index);
    }
    this.setState({ paginationIndex: index });

    if (onMomentumScrollEnd) {
      onMomentumScrollEnd();
    }
  };

  _onScrollToIndexFailed = info => {
    setTimeout(() => this._scrollToIndex(info.index, false));
  };

  _keyExtractor = (item, index) => index;

  renderChildren = ({ item }) => item;

  render() {
    const {
      vertical,
      showPagination,
      children,
      PaginationComponent,
      paginationActiveColor,
      paginationDefaultColor,
      ...props
    } = this.props;

    const flatListProps = {
      ref: component => {
        this.flatListRef = component;
      },
      keyExtractor: this._keyExtractor,
      horizontal: !vertical,
      showsHorizontalScrollIndicator: false,
      showsVerticalScrollIndicator: false,
      pagingEnabled: true,
      onMomentumScrollEnd: this._onMomentumScrollEnd,
      onScrollToIndexFailed: this._onScrollToIndexFailed,
      ...props,
      data: this._data,
      renderItem: this._renderItem,
      initialNumToRender: this._initialNumToRender
      // inverted
      // onViewableItemsChanged={(data, index, x) => {
      // getItemLayout={(data, index, x) => {
      // initialScrollIndex={this.state.index}
      // ListFooterComponent=
      // ListEmptyComponent loading...
    };

    const paginationProps = {
      data: this._data,
      paginationIndex: this.state.paginationIndex,
      scrollToIndex: this._scrollToIndex,
      paginationActiveColor,
      paginationDefaultColor
    };

    return (
      <View>
        <FlatList {...flatListProps} />
        {showPagination && <PaginationComponent {...paginationProps} />}
      </View>
    );
  }
}
