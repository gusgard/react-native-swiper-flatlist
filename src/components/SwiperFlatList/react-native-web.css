/* Fix for height https://github.com/necolas/react-native-web/issues/1299#issuecomment-717428165 */
[data-paging-enabled-fix] > div > div > div {
  height: 100%;
}
