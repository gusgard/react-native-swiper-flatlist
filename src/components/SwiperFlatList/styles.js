import { StyleSheet } from 'react-native';

import { vertical, colors, horizontal, width } from '../../themes';

export default StyleSheet.create({
  paginationContainer: {
    position: 'absolute',
    flexDirection: 'row',
    marginVertical: vertical.xxSmall,
    justifyContent: 'center',
    bottom: 0,
    left: width * 0.25,
    right: width * 0.25
  },
  pagination: {
    width: horizontal.small,
    height: horizontal.small,
    backgroundColor: colors.gray,
    borderRadius: 25,
    marginHorizontal: horizontal.xSmall
  },
  paginationActive: {
    backgroundColor: colors.primaryDark
  }
});
