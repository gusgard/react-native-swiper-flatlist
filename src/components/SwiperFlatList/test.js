import React from 'react';
import { Image } from 'react-native';
import { render } from 'react-native-testing-library';

import SwiperFlatList from './SwiperFlatList';

const logo = { uri: 'https://...' };
const items = [{ id: 1, thumbnail: logo }, { id: 2, thumbnail: logo }];

describe('swiper flatlist', () => {
  it('renders correctly', () => {
    const { toJSON } = render(
      <SwiperFlatList renderItem={({ item }) => <Image source={item.thumbnail} />} data={items} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
  it('renders children', () => {
    const { toJSON } = render(
      <SwiperFlatList>
        <Image source={items[0].thumbnail} />
        <Image source={items[1].thumbnail} />
      </SwiperFlatList>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
