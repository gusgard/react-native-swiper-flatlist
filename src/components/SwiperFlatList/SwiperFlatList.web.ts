// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React from 'react';
const { SwiperFlatList: SwiperFlatListBase } = require('./SwiperFlatList.tsx');

try {
  require('./global.css');
} catch (error) {}

export const SwiperFlatList = SwiperFlatListBase;
