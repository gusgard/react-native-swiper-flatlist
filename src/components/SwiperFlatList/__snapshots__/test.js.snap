// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swiper flatlist renders children 1`] = `
<View>
  <FlatList
    autoplay={false}
    autoplayDelay={3}
    autoplayDirection={true}
    autoplayLoop={false}
    data={
      Array [
        <Image
          source={
            Object {
              "uri": "https://...",
            }
          }
        />,
        <Image
          source={
            Object {
              "uri": "https://...",
            }
          }
        />,
      ]
    }
    disableVirtualization={false}
    horizontal={true}
    index={0}
    initialNumToRender={1}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    onMomentumScrollEnd={[Function]}
    onScrollToIndexFailed={[Function]}
    pagingEnabled={true}
    renderAll={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`swiper flatlist renders correctly 1`] = `
<View>
  <FlatList
    autoplay={false}
    autoplayDelay={3}
    autoplayDirection={true}
    autoplayLoop={false}
    data={
      Array [
        Object {
          "id": 1,
          "thumbnail": Object {
            "uri": "https://...",
          },
        },
        Object {
          "id": 2,
          "thumbnail": Object {
            "uri": "https://...",
          },
        },
      ]
    }
    disableVirtualization={false}
    horizontal={true}
    index={0}
    initialNumToRender={1}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    onMomentumScrollEnd={[Function]}
    onScrollToIndexFailed={[Function]}
    pagingEnabled={true}
    renderAll={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;
